lib/dispatch-rider/publisher/base.rb

```diff
@@ -1,37 +1,46 @@
 require 'securerandom'
 
 module DispatchRider
+  # Main template for a dispatch rider publisher.
   class Publisher::Base
+    include Callbacks::Support
 
     class << self
-
+      # @param [Symbol] subject
       def subject(subject)
         @subject = subject
       end
 
+      # @param [Array<Symbol>, Symbol] destinations
       def destinations(destinations)
         @destinations = Array(destinations)
       end
 
+      # @return [DispatchRider::Publisher]
       def default_publisher
         @@default_publisher ||= DispatchRider::Publisher.new
       end
 
       def publish(*args, &block)
-        raise NotImplementedError
+        raise NotImplementedError, "subclass of DispatchRider::Publisher::Base must implement .publish"
       end
-
     end
 
     def initialize(publisher = nil)
       @publisher = publisher
     end
 
+    # @param [Hash] body
     def publish(body)
       raise ArgumentError, 'body should be a hash' unless body.kind_of?(Hash)
-      publisher.publish(destinations: destinations, message: { subject: subject, body: body })
+
+      callbacks.invoke(:publish, body) do
+        publisher.publish(destinations: destinations, message: { subject: subject, body: body })
```

Comment:
Hmm my only concern is that this needs to be duplicated wherever you do publishing .. 

The self publishing support code that's not in the gem yet, the map/reduce code we have .. they all use publisher directly, not Publisher::Base, and the whole trend of having a dedicated second class is dying.

Perhaps this functionality should be factored out into composable modules, so that Publisher::Base is a convenience base class, but you can do all the same by manually including its parts.